{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "da52a68a",
   "metadata": {},
   "source": [
    "<img src=\"https://heig-vd.ch/docs/default-source/doc-global-newsletter/2020-slim.svg\" alt=\"Logo HEIG-VD\" style=\"width: 80px;\" align=\"right\"/>\n",
    "\n",
    "# Cours APN - Labo 2 : Comparaison de méthodes <br>de groupement sur des données biomédicales"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b776ea11",
   "metadata": {},
   "source": [
    "## Résumé\n",
    "\n",
    "L'objectif de ce laboratoire est de comparer deux méthodes de groupement : _k_-moyennes et DBSCAN.  \n",
    "\n",
    "Le jeu de données vient du domaine biomédical et possède une annotation de référence.  \n",
    "\n",
    "Les tâches demandées sont les suivantes :\n",
    "* préparer les données à partir de données brutes disponibles en ligne\n",
    "* pour chaque méthode de groupement\n",
    "  - décider quels sont les meilleurs hyper-paramètres par évaluation intrinsèque\n",
    "  - avec ces paramètres, afficher aussi les scores par évaluation extrinsèque\n",
    "  - visualiser les groupements en 2D\n",
    "* comparer les deux méthodes de groupement et conclure."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "29e26888",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Nom et prénom : "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "c9e1ef26",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "import os\n",
    "os.environ[\"OMP_NUM_THREADS\"] = \"1\" # KMeans is known to have a memory leak ... set OMP_NUM_THREADS=1\n",
    "\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.decomposition import PCA\n",
    "\n",
    "from sklearn.cluster import KMeans\n",
    "from sklearn.cluster import DBSCAN\n",
    "\n",
    "from sklearn.metrics import davies_bouldin_score as dbs\n",
    "from sklearn.metrics import silhouette_score as sis\n",
    "from sklearn.metrics import adjusted_rand_score as ars\n",
    "from sklearn.metrics import f1_score\n",
    "\n",
    "from matplotlib.colors import Normalize # peut aider au coloriage des groupes (ou utiliser Seaborn)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee11ccc6",
   "metadata": {},
   "source": [
    "## 1. Prise en main et analyse exploratoire des données\n",
    "\n",
    "Les données biomédicales ont été présentées dans l'article de Todd R. Golub et al., [Molecular classification of cancer: class discovery and class prediction by gene expression monitoring](https://www.science.org/doi/abs/10.1126/science.286.5439.531), *Science*, 286:531-537, 1999.  Ces données représentent l'expression de **7129 gènes** dans des échantillons sanguins et de moëlle osseuse provenant de 72 patients souffrant de leucémie.  Pour chaque gène, son niveau d'expression est codé comme un nombre entier, positif ou négatif.  Le type de leucémie a été annoté pour chaque patient comme ALL ou AML (*acute lymphoblastic leukemia* ou *acute myeloid leukemia*) : ce sont les deux classes de référence.\n",
    "\n",
    "Les données sont à télécharger depuis [une version fournie par C. Crawford sur Kaggle](https://www.kaggle.com/datasets/crawford/gene-expression) (pour information, une autre version se trouve dans le [package R golubEsets](https://master.bioconductor.org/packages/release/data/experiment/html/golubEsets.html)).  Si vous n'avez pas accès à Kaggle, le fichier `archive.zip` est fourni avec ce notebook.  Les données brutes contiennent :\n",
    "* un fichier `train` avec les données de 38 patients : noms des gènes (longs et courts), niveau d'expression, signifiance ('call')\n",
    "* un fichier `independent` avec les données de 34 autres patients\n",
    "* un fichier `actual` qui indique pour chaque patient son type de leucémie (ALL ou AML)\n",
    "\n",
    "**Le premier but est d'obtenir une DataFrame avec les 72 patients (lignes) et les niveaux des 7129 gènes (colonnes).  Le numéro de patient sert d'index.  Enfin, une colonne indiquera le type de chaque patient.**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c79da7d",
   "metadata": {},
   "source": [
    "a. Veuillez compléter le code ci-dessous pour transformer le premier fichier de données en un tableau:\n",
    "* avec 7129 colonnes correspondant à la valeur exprimée de chaque gène pour chaque patient\n",
    "* les noms des 7129 colonnes sont les noms courts de chaque gène (connus comme *Gene Accession Number*) \n",
    "* indexé par le numéro des patients, qui sera appelé 'Patient'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "7d0c4cf4",
   "metadata": {},
   "outputs": [],
   "source": [
    "actual = pd.read_csv(\"archive/actual.csv\")\n",
    "independent = pd.read_csv(\"archive/data_set_ALL_AML_independent.csv\")\n",
    "train = pd.read_csv(\"archive/data_set_ALL_AML_train.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "99459991",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 7129 entries, 0 to 7128\n",
      "Data columns (total 78 columns):\n",
      " #   Column                 Non-Null Count  Dtype \n",
      "---  ------                 --------------  ----- \n",
      " 0   Gene Description       7129 non-null   object\n",
      " 1   Gene Accession Number  7129 non-null   object\n",
      " 2   1                      7129 non-null   int64 \n",
      " 3   call                   7129 non-null   object\n",
      " 4   2                      7129 non-null   int64 \n",
      " 5   call.1                 7129 non-null   object\n",
      " 6   3                      7129 non-null   int64 \n",
      " 7   call.2                 7129 non-null   object\n",
      " 8   4                      7129 non-null   int64 \n",
      " 9   call.3                 7129 non-null   object\n",
      " 10  5                      7129 non-null   int64 \n",
      " 11  call.4                 7129 non-null   object\n",
      " 12  6                      7129 non-null   int64 \n",
      " 13  call.5                 7129 non-null   object\n",
      " 14  7                      7129 non-null   int64 \n",
      " 15  call.6                 7129 non-null   object\n",
      " 16  8                      7129 non-null   int64 \n",
      " 17  call.7                 7129 non-null   object\n",
      " 18  9                      7129 non-null   int64 \n",
      " 19  call.8                 7129 non-null   object\n",
      " 20  10                     7129 non-null   int64 \n",
      " 21  call.9                 7129 non-null   object\n",
      " 22  11                     7129 non-null   int64 \n",
      " 23  call.10                7129 non-null   object\n",
      " 24  12                     7129 non-null   int64 \n",
      " 25  call.11                7129 non-null   object\n",
      " 26  13                     7129 non-null   int64 \n",
      " 27  call.12                7129 non-null   object\n",
      " 28  14                     7129 non-null   int64 \n",
      " 29  call.13                7129 non-null   object\n",
      " 30  15                     7129 non-null   int64 \n",
      " 31  call.14                7129 non-null   object\n",
      " 32  16                     7129 non-null   int64 \n",
      " 33  call.15                7129 non-null   object\n",
      " 34  17                     7129 non-null   int64 \n",
      " 35  call.16                7129 non-null   object\n",
      " 36  18                     7129 non-null   int64 \n",
      " 37  call.17                7129 non-null   object\n",
      " 38  19                     7129 non-null   int64 \n",
      " 39  call.18                7129 non-null   object\n",
      " 40  20                     7129 non-null   int64 \n",
      " 41  call.19                7129 non-null   object\n",
      " 42  21                     7129 non-null   int64 \n",
      " 43  call.20                7129 non-null   object\n",
      " 44  22                     7129 non-null   int64 \n",
      " 45  call.21                7129 non-null   object\n",
      " 46  23                     7129 non-null   int64 \n",
      " 47  call.22                7129 non-null   object\n",
      " 48  24                     7129 non-null   int64 \n",
      " 49  call.23                7129 non-null   object\n",
      " 50  25                     7129 non-null   int64 \n",
      " 51  call.24                7129 non-null   object\n",
      " 52  26                     7129 non-null   int64 \n",
      " 53  call.25                7129 non-null   object\n",
      " 54  27                     7129 non-null   int64 \n",
      " 55  call.26                7129 non-null   object\n",
      " 56  34                     7129 non-null   int64 \n",
      " 57  call.27                7129 non-null   object\n",
      " 58  35                     7129 non-null   int64 \n",
      " 59  call.28                7129 non-null   object\n",
      " 60  36                     7129 non-null   int64 \n",
      " 61  call.29                7129 non-null   object\n",
      " 62  37                     7129 non-null   int64 \n",
      " 63  call.30                7129 non-null   object\n",
      " 64  38                     7129 non-null   int64 \n",
      " 65  call.31                7129 non-null   object\n",
      " 66  28                     7129 non-null   int64 \n",
      " 67  call.32                7129 non-null   object\n",
      " 68  29                     7129 non-null   int64 \n",
      " 69  call.33                7129 non-null   object\n",
      " 70  30                     7129 non-null   int64 \n",
      " 71  call.34                7129 non-null   object\n",
      " 72  31                     7129 non-null   int64 \n",
      " 73  call.35                7129 non-null   object\n",
      " 74  32                     7129 non-null   int64 \n",
      " 75  call.36                7129 non-null   object\n",
      " 76  33                     7129 non-null   int64 \n",
      " 77  call.37                7129 non-null   object\n",
      "dtypes: int64(38), object(40)\n",
      "memory usage: 4.2+ MB\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "39712166",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = train.iloc[:, 2:].select_dtypes(include=[\"int64\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "f8353d52",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>...</th>\n",
       "      <th>35</th>\n",
       "      <th>36</th>\n",
       "      <th>37</th>\n",
       "      <th>38</th>\n",
       "      <th>28</th>\n",
       "      <th>29</th>\n",
       "      <th>30</th>\n",
       "      <th>31</th>\n",
       "      <th>32</th>\n",
       "      <th>33</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-214</td>\n",
       "      <td>-139</td>\n",
       "      <td>-76</td>\n",
       "      <td>-135</td>\n",
       "      <td>-106</td>\n",
       "      <td>-138</td>\n",
       "      <td>-72</td>\n",
       "      <td>-413</td>\n",
       "      <td>5</td>\n",
       "      <td>-88</td>\n",
       "      <td>...</td>\n",
       "      <td>7</td>\n",
       "      <td>-213</td>\n",
       "      <td>-25</td>\n",
       "      <td>-72</td>\n",
       "      <td>-4</td>\n",
       "      <td>15</td>\n",
       "      <td>-318</td>\n",
       "      <td>-32</td>\n",
       "      <td>-124</td>\n",
       "      <td>-135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-153</td>\n",
       "      <td>-73</td>\n",
       "      <td>-49</td>\n",
       "      <td>-114</td>\n",
       "      <td>-125</td>\n",
       "      <td>-85</td>\n",
       "      <td>-144</td>\n",
       "      <td>-260</td>\n",
       "      <td>-127</td>\n",
       "      <td>-105</td>\n",
       "      <td>...</td>\n",
       "      <td>-100</td>\n",
       "      <td>-252</td>\n",
       "      <td>-20</td>\n",
       "      <td>-139</td>\n",
       "      <td>-116</td>\n",
       "      <td>-114</td>\n",
       "      <td>-192</td>\n",
       "      <td>-49</td>\n",
       "      <td>-79</td>\n",
       "      <td>-186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>-58</td>\n",
       "      <td>-1</td>\n",
       "      <td>-307</td>\n",
       "      <td>265</td>\n",
       "      <td>-76</td>\n",
       "      <td>215</td>\n",
       "      <td>238</td>\n",
       "      <td>7</td>\n",
       "      <td>106</td>\n",
       "      <td>42</td>\n",
       "      <td>...</td>\n",
       "      <td>-57</td>\n",
       "      <td>136</td>\n",
       "      <td>124</td>\n",
       "      <td>-1</td>\n",
       "      <td>-125</td>\n",
       "      <td>2</td>\n",
       "      <td>-95</td>\n",
       "      <td>49</td>\n",
       "      <td>-37</td>\n",
       "      <td>-70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>88</td>\n",
       "      <td>283</td>\n",
       "      <td>309</td>\n",
       "      <td>12</td>\n",
       "      <td>168</td>\n",
       "      <td>71</td>\n",
       "      <td>55</td>\n",
       "      <td>-2</td>\n",
       "      <td>268</td>\n",
       "      <td>219</td>\n",
       "      <td>...</td>\n",
       "      <td>132</td>\n",
       "      <td>318</td>\n",
       "      <td>325</td>\n",
       "      <td>392</td>\n",
       "      <td>241</td>\n",
       "      <td>193</td>\n",
       "      <td>312</td>\n",
       "      <td>230</td>\n",
       "      <td>330</td>\n",
       "      <td>337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-295</td>\n",
       "      <td>-264</td>\n",
       "      <td>-376</td>\n",
       "      <td>-419</td>\n",
       "      <td>-230</td>\n",
       "      <td>-272</td>\n",
       "      <td>-399</td>\n",
       "      <td>-541</td>\n",
       "      <td>-210</td>\n",
       "      <td>-178</td>\n",
       "      <td>...</td>\n",
       "      <td>-377</td>\n",
       "      <td>-209</td>\n",
       "      <td>-396</td>\n",
       "      <td>-324</td>\n",
       "      <td>-191</td>\n",
       "      <td>-51</td>\n",
       "      <td>-139</td>\n",
       "      <td>-367</td>\n",
       "      <td>-188</td>\n",
       "      <td>-407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7124</th>\n",
       "      <td>793</td>\n",
       "      <td>782</td>\n",
       "      <td>1138</td>\n",
       "      <td>627</td>\n",
       "      <td>250</td>\n",
       "      <td>645</td>\n",
       "      <td>1140</td>\n",
       "      <td>1799</td>\n",
       "      <td>758</td>\n",
       "      <td>570</td>\n",
       "      <td>...</td>\n",
       "      <td>752</td>\n",
       "      <td>1293</td>\n",
       "      <td>1733</td>\n",
       "      <td>1567</td>\n",
       "      <td>987</td>\n",
       "      <td>279</td>\n",
       "      <td>737</td>\n",
       "      <td>588</td>\n",
       "      <td>1170</td>\n",
       "      <td>2315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7125</th>\n",
       "      <td>329</td>\n",
       "      <td>295</td>\n",
       "      <td>777</td>\n",
       "      <td>170</td>\n",
       "      <td>314</td>\n",
       "      <td>341</td>\n",
       "      <td>482</td>\n",
       "      <td>446</td>\n",
       "      <td>385</td>\n",
       "      <td>359</td>\n",
       "      <td>...</td>\n",
       "      <td>295</td>\n",
       "      <td>342</td>\n",
       "      <td>304</td>\n",
       "      <td>627</td>\n",
       "      <td>279</td>\n",
       "      <td>51</td>\n",
       "      <td>227</td>\n",
       "      <td>361</td>\n",
       "      <td>284</td>\n",
       "      <td>250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7126</th>\n",
       "      <td>36</td>\n",
       "      <td>11</td>\n",
       "      <td>41</td>\n",
       "      <td>-50</td>\n",
       "      <td>14</td>\n",
       "      <td>26</td>\n",
       "      <td>10</td>\n",
       "      <td>59</td>\n",
       "      <td>115</td>\n",
       "      <td>9</td>\n",
       "      <td>...</td>\n",
       "      <td>28</td>\n",
       "      <td>26</td>\n",
       "      <td>12</td>\n",
       "      <td>21</td>\n",
       "      <td>22</td>\n",
       "      <td>6</td>\n",
       "      <td>-9</td>\n",
       "      <td>-26</td>\n",
       "      <td>39</td>\n",
       "      <td>-12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7127</th>\n",
       "      <td>191</td>\n",
       "      <td>76</td>\n",
       "      <td>228</td>\n",
       "      <td>126</td>\n",
       "      <td>56</td>\n",
       "      <td>193</td>\n",
       "      <td>369</td>\n",
       "      <td>781</td>\n",
       "      <td>244</td>\n",
       "      <td>171</td>\n",
       "      <td>...</td>\n",
       "      <td>1558</td>\n",
       "      <td>246</td>\n",
       "      <td>3193</td>\n",
       "      <td>2520</td>\n",
       "      <td>662</td>\n",
       "      <td>2484</td>\n",
       "      <td>371</td>\n",
       "      <td>133</td>\n",
       "      <td>298</td>\n",
       "      <td>790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7128</th>\n",
       "      <td>-37</td>\n",
       "      <td>-14</td>\n",
       "      <td>-41</td>\n",
       "      <td>-91</td>\n",
       "      <td>-25</td>\n",
       "      <td>-53</td>\n",
       "      <td>-42</td>\n",
       "      <td>20</td>\n",
       "      <td>-39</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>-67</td>\n",
       "      <td>23</td>\n",
       "      <td>-33</td>\n",
       "      <td>0</td>\n",
       "      <td>-46</td>\n",
       "      <td>-2</td>\n",
       "      <td>-31</td>\n",
       "      <td>-32</td>\n",
       "      <td>-3</td>\n",
       "      <td>-10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7129 rows × 38 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        1    2     3    4    5    6     7     8    9   10  ...    35    36  \\\n",
       "0    -214 -139   -76 -135 -106 -138   -72  -413    5  -88  ...     7  -213   \n",
       "1    -153  -73   -49 -114 -125  -85  -144  -260 -127 -105  ...  -100  -252   \n",
       "2     -58   -1  -307  265  -76  215   238     7  106   42  ...   -57   136   \n",
       "3      88  283   309   12  168   71    55    -2  268  219  ...   132   318   \n",
       "4    -295 -264  -376 -419 -230 -272  -399  -541 -210 -178  ...  -377  -209   \n",
       "...   ...  ...   ...  ...  ...  ...   ...   ...  ...  ...  ...   ...   ...   \n",
       "7124  793  782  1138  627  250  645  1140  1799  758  570  ...   752  1293   \n",
       "7125  329  295   777  170  314  341   482   446  385  359  ...   295   342   \n",
       "7126   36   11    41  -50   14   26    10    59  115    9  ...    28    26   \n",
       "7127  191   76   228  126   56  193   369   781  244  171  ...  1558   246   \n",
       "7128  -37  -14   -41  -91  -25  -53   -42    20  -39    7  ...   -67    23   \n",
       "\n",
       "        37    38   28    29   30   31    32    33  \n",
       "0      -25   -72   -4    15 -318  -32  -124  -135  \n",
       "1      -20  -139 -116  -114 -192  -49   -79  -186  \n",
       "2      124    -1 -125     2  -95   49   -37   -70  \n",
       "3      325   392  241   193  312  230   330   337  \n",
       "4     -396  -324 -191   -51 -139 -367  -188  -407  \n",
       "...    ...   ...  ...   ...  ...  ...   ...   ...  \n",
       "7124  1733  1567  987   279  737  588  1170  2315  \n",
       "7125   304   627  279    51  227  361   284   250  \n",
       "7126    12    21   22     6   -9  -26    39   -12  \n",
       "7127  3193  2520  662  2484  371  133   298   790  \n",
       "7128   -33     0  -46    -2  -31  -32    -3   -10  \n",
       "\n",
       "[7129 rows x 38 columns]"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "af9f7aab",
   "metadata": {},
   "outputs": [],
   "source": [
    "def conversion_donnees_brutes(path : str) -> pd.DataFrame:\n",
    "    # Retourne la DataFrame selon la consigne.\n",
    "    # Code à compléter :\n",
    "    # Retourne la DataFrame selon la consigne.\n",
    "    # Code à compléter :\n",
    "\n",
    "    # Importer les données\n",
    "    data = pd.read_csv(path)\n",
    "\n",
    "    # Extraire les colonnes de gènes\n",
    "    genes = data.iloc[:, 2:].select_dtypes(include=[\"int64\"])\n",
    "\n",
    "    # Transposer les données des gènes pour que chaque colonne corresponde à un \"Gene Accession Number\"\n",
    "    genes_transposed = genes.T\n",
    "\n",
    "    # Attribuer les noms des colonnes en utilisant \"Gene Accession Number\"\n",
    "    genes_transposed.columns = data[\"Gene Accession Number\"]\n",
    "\n",
    "    return genes_transposed\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "282eeb90",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(38, 7129)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Gene Accession Number</th>\n",
       "      <th>AFFX-BioB-5_at</th>\n",
       "      <th>AFFX-BioB-M_at</th>\n",
       "      <th>AFFX-BioB-3_at</th>\n",
       "      <th>AFFX-BioC-5_at</th>\n",
       "      <th>AFFX-BioC-3_at</th>\n",
       "      <th>AFFX-BioDn-5_at</th>\n",
       "      <th>AFFX-BioDn-3_at</th>\n",
       "      <th>AFFX-CreX-5_at</th>\n",
       "      <th>AFFX-CreX-3_at</th>\n",
       "      <th>AFFX-BioB-5_st</th>\n",
       "      <th>...</th>\n",
       "      <th>U48730_at</th>\n",
       "      <th>U58516_at</th>\n",
       "      <th>U73738_at</th>\n",
       "      <th>X06956_at</th>\n",
       "      <th>X16699_at</th>\n",
       "      <th>X83863_at</th>\n",
       "      <th>Z17240_at</th>\n",
       "      <th>L49218_f_at</th>\n",
       "      <th>M71243_f_at</th>\n",
       "      <th>Z78285_f_at</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-214</td>\n",
       "      <td>-153</td>\n",
       "      <td>-58</td>\n",
       "      <td>88</td>\n",
       "      <td>-295</td>\n",
       "      <td>-558</td>\n",
       "      <td>199</td>\n",
       "      <td>-176</td>\n",
       "      <td>252</td>\n",
       "      <td>206</td>\n",
       "      <td>...</td>\n",
       "      <td>185</td>\n",
       "      <td>511</td>\n",
       "      <td>-125</td>\n",
       "      <td>389</td>\n",
       "      <td>-37</td>\n",
       "      <td>793</td>\n",
       "      <td>329</td>\n",
       "      <td>36</td>\n",
       "      <td>191</td>\n",
       "      <td>-37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>-139</td>\n",
       "      <td>-73</td>\n",
       "      <td>-1</td>\n",
       "      <td>283</td>\n",
       "      <td>-264</td>\n",
       "      <td>-400</td>\n",
       "      <td>-330</td>\n",
       "      <td>-168</td>\n",
       "      <td>101</td>\n",
       "      <td>74</td>\n",
       "      <td>...</td>\n",
       "      <td>169</td>\n",
       "      <td>837</td>\n",
       "      <td>-36</td>\n",
       "      <td>442</td>\n",
       "      <td>-17</td>\n",
       "      <td>782</td>\n",
       "      <td>295</td>\n",
       "      <td>11</td>\n",
       "      <td>76</td>\n",
       "      <td>-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-76</td>\n",
       "      <td>-49</td>\n",
       "      <td>-307</td>\n",
       "      <td>309</td>\n",
       "      <td>-376</td>\n",
       "      <td>-650</td>\n",
       "      <td>33</td>\n",
       "      <td>-367</td>\n",
       "      <td>206</td>\n",
       "      <td>-215</td>\n",
       "      <td>...</td>\n",
       "      <td>315</td>\n",
       "      <td>1199</td>\n",
       "      <td>33</td>\n",
       "      <td>168</td>\n",
       "      <td>52</td>\n",
       "      <td>1138</td>\n",
       "      <td>777</td>\n",
       "      <td>41</td>\n",
       "      <td>228</td>\n",
       "      <td>-41</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 7129 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene Accession Number  AFFX-BioB-5_at  AFFX-BioB-M_at  AFFX-BioB-3_at  \\\n",
       "1                                -214            -153             -58   \n",
       "2                                -139             -73              -1   \n",
       "3                                 -76             -49            -307   \n",
       "\n",
       "Gene Accession Number  AFFX-BioC-5_at  AFFX-BioC-3_at  AFFX-BioDn-5_at  \\\n",
       "1                                  88            -295             -558   \n",
       "2                                 283            -264             -400   \n",
       "3                                 309            -376             -650   \n",
       "\n",
       "Gene Accession Number  AFFX-BioDn-3_at  AFFX-CreX-5_at  AFFX-CreX-3_at  \\\n",
       "1                                  199            -176             252   \n",
       "2                                 -330            -168             101   \n",
       "3                                   33            -367             206   \n",
       "\n",
       "Gene Accession Number  AFFX-BioB-5_st  ...  U48730_at  U58516_at  U73738_at  \\\n",
       "1                                 206  ...        185        511       -125   \n",
       "2                                  74  ...        169        837        -36   \n",
       "3                                -215  ...        315       1199         33   \n",
       "\n",
       "Gene Accession Number  X06956_at  X16699_at  X83863_at  Z17240_at  \\\n",
       "1                            389        -37        793        329   \n",
       "2                            442        -17        782        295   \n",
       "3                            168         52       1138        777   \n",
       "\n",
       "Gene Accession Number  L49218_f_at  M71243_f_at  Z78285_f_at  \n",
       "1                               36          191          -37  \n",
       "2                               11           76          -14  \n",
       "3                               41          228          -41  \n",
       "\n",
       "[3 rows x 7129 columns]"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_df = conversion_donnees_brutes(os.path.join('archive', 'data_set_ALL_AML_train.csv'))\n",
    "print(train_df.shape) # vérifie le résultat : (38, 7129)\n",
    "train_df.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c63d0b8b",
   "metadata": {},
   "source": [
    "b. Veuillez appliquer la même fonction au deuxième fichier de données."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "dbc82e7f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(34, 7129)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Gene Accession Number</th>\n",
       "      <th>AFFX-BioB-5_at</th>\n",
       "      <th>AFFX-BioB-M_at</th>\n",
       "      <th>AFFX-BioB-3_at</th>\n",
       "      <th>AFFX-BioC-5_at</th>\n",
       "      <th>AFFX-BioC-3_at</th>\n",
       "      <th>AFFX-BioDn-5_at</th>\n",
       "      <th>AFFX-BioDn-3_at</th>\n",
       "      <th>AFFX-CreX-5_at</th>\n",
       "      <th>AFFX-CreX-3_at</th>\n",
       "      <th>AFFX-BioB-5_st</th>\n",
       "      <th>...</th>\n",
       "      <th>U48730_at</th>\n",
       "      <th>U58516_at</th>\n",
       "      <th>U73738_at</th>\n",
       "      <th>X06956_at</th>\n",
       "      <th>X16699_at</th>\n",
       "      <th>X83863_at</th>\n",
       "      <th>Z17240_at</th>\n",
       "      <th>L49218_f_at</th>\n",
       "      <th>M71243_f_at</th>\n",
       "      <th>Z78285_f_at</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>-342</td>\n",
       "      <td>-200</td>\n",
       "      <td>41</td>\n",
       "      <td>328</td>\n",
       "      <td>-224</td>\n",
       "      <td>-427</td>\n",
       "      <td>-656</td>\n",
       "      <td>-292</td>\n",
       "      <td>137</td>\n",
       "      <td>-144</td>\n",
       "      <td>...</td>\n",
       "      <td>277</td>\n",
       "      <td>1023</td>\n",
       "      <td>67</td>\n",
       "      <td>214</td>\n",
       "      <td>-135</td>\n",
       "      <td>1074</td>\n",
       "      <td>475</td>\n",
       "      <td>48</td>\n",
       "      <td>168</td>\n",
       "      <td>-70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>-87</td>\n",
       "      <td>-248</td>\n",
       "      <td>262</td>\n",
       "      <td>295</td>\n",
       "      <td>-226</td>\n",
       "      <td>-493</td>\n",
       "      <td>367</td>\n",
       "      <td>-452</td>\n",
       "      <td>194</td>\n",
       "      <td>162</td>\n",
       "      <td>...</td>\n",
       "      <td>83</td>\n",
       "      <td>529</td>\n",
       "      <td>-295</td>\n",
       "      <td>352</td>\n",
       "      <td>-67</td>\n",
       "      <td>67</td>\n",
       "      <td>263</td>\n",
       "      <td>-33</td>\n",
       "      <td>-33</td>\n",
       "      <td>-21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>22</td>\n",
       "      <td>-153</td>\n",
       "      <td>17</td>\n",
       "      <td>276</td>\n",
       "      <td>-211</td>\n",
       "      <td>-250</td>\n",
       "      <td>55</td>\n",
       "      <td>-141</td>\n",
       "      <td>0</td>\n",
       "      <td>500</td>\n",
       "      <td>...</td>\n",
       "      <td>413</td>\n",
       "      <td>399</td>\n",
       "      <td>16</td>\n",
       "      <td>558</td>\n",
       "      <td>24</td>\n",
       "      <td>893</td>\n",
       "      <td>297</td>\n",
       "      <td>6</td>\n",
       "      <td>1971</td>\n",
       "      <td>-42</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 7129 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene Accession Number  AFFX-BioB-5_at  AFFX-BioB-M_at  AFFX-BioB-3_at  \\\n",
       "39                               -342            -200              41   \n",
       "40                                -87            -248             262   \n",
       "42                                 22            -153              17   \n",
       "\n",
       "Gene Accession Number  AFFX-BioC-5_at  AFFX-BioC-3_at  AFFX-BioDn-5_at  \\\n",
       "39                                328            -224             -427   \n",
       "40                                295            -226             -493   \n",
       "42                                276            -211             -250   \n",
       "\n",
       "Gene Accession Number  AFFX-BioDn-3_at  AFFX-CreX-5_at  AFFX-CreX-3_at  \\\n",
       "39                                -656            -292             137   \n",
       "40                                 367            -452             194   \n",
       "42                                  55            -141               0   \n",
       "\n",
       "Gene Accession Number  AFFX-BioB-5_st  ...  U48730_at  U58516_at  U73738_at  \\\n",
       "39                               -144  ...        277       1023         67   \n",
       "40                                162  ...         83        529       -295   \n",
       "42                                500  ...        413        399         16   \n",
       "\n",
       "Gene Accession Number  X06956_at  X16699_at  X83863_at  Z17240_at  \\\n",
       "39                           214       -135       1074        475   \n",
       "40                           352        -67         67        263   \n",
       "42                           558         24        893        297   \n",
       "\n",
       "Gene Accession Number  L49218_f_at  M71243_f_at  Z78285_f_at  \n",
       "39                              48          168          -70  \n",
       "40                             -33          -33          -21  \n",
       "42                               6         1971          -42  \n",
       "\n",
       "[3 rows x 7129 columns]"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_df = conversion_donnees_brutes(os.path.join('archive', 'data_set_ALL_AML_independent.csv'))\n",
    "print(test_df.shape) # vérifier le résultat : (34, 7129)\n",
    "test_df.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a071d50b",
   "metadata": {},
   "source": [
    "c. Veuillez fusionner les deux data frames en mettant les patients dans l'ordre 1 à 72."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "df43634a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Gene Accession Number</th>\n",
       "      <th>A28102_at</th>\n",
       "      <th>AB000114_at</th>\n",
       "      <th>AB000115_at</th>\n",
       "      <th>AB000220_at</th>\n",
       "      <th>AB000381_s_at</th>\n",
       "      <th>AB000409_at</th>\n",
       "      <th>AB000410_s_at</th>\n",
       "      <th>AB000449_at</th>\n",
       "      <th>AB000450_at</th>\n",
       "      <th>AB000460_at</th>\n",
       "      <th>...</th>\n",
       "      <th>Z84721_cds2_at</th>\n",
       "      <th>Z84722_at</th>\n",
       "      <th>Z86000_at</th>\n",
       "      <th>Z93784_at</th>\n",
       "      <th>Z94753_s_at</th>\n",
       "      <th>Z95624_at</th>\n",
       "      <th>Z96810_at</th>\n",
       "      <th>Z97054_xpt2_at</th>\n",
       "      <th>Z97074_at</th>\n",
       "      <th>hum_alu_at</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>151</td>\n",
       "      <td>72</td>\n",
       "      <td>281</td>\n",
       "      <td>36</td>\n",
       "      <td>29</td>\n",
       "      <td>-299</td>\n",
       "      <td>-336</td>\n",
       "      <td>57</td>\n",
       "      <td>186</td>\n",
       "      <td>1647</td>\n",
       "      <td>...</td>\n",
       "      <td>8656</td>\n",
       "      <td>-156</td>\n",
       "      <td>538</td>\n",
       "      <td>434</td>\n",
       "      <td>168</td>\n",
       "      <td>116</td>\n",
       "      <td>-122</td>\n",
       "      <td>398</td>\n",
       "      <td>188</td>\n",
       "      <td>15091</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>263</td>\n",
       "      <td>21</td>\n",
       "      <td>250</td>\n",
       "      <td>43</td>\n",
       "      <td>8</td>\n",
       "      <td>-103</td>\n",
       "      <td>-361</td>\n",
       "      <td>169</td>\n",
       "      <td>219</td>\n",
       "      <td>2043</td>\n",
       "      <td>...</td>\n",
       "      <td>13285</td>\n",
       "      <td>-241</td>\n",
       "      <td>863</td>\n",
       "      <td>506</td>\n",
       "      <td>70</td>\n",
       "      <td>94</td>\n",
       "      <td>-40</td>\n",
       "      <td>649</td>\n",
       "      <td>147</td>\n",
       "      <td>11038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>88</td>\n",
       "      <td>-27</td>\n",
       "      <td>358</td>\n",
       "      <td>42</td>\n",
       "      <td>11</td>\n",
       "      <td>142</td>\n",
       "      <td>-508</td>\n",
       "      <td>359</td>\n",
       "      <td>237</td>\n",
       "      <td>1997</td>\n",
       "      <td>...</td>\n",
       "      <td>15704</td>\n",
       "      <td>-581</td>\n",
       "      <td>737</td>\n",
       "      <td>770</td>\n",
       "      <td>141</td>\n",
       "      <td>57</td>\n",
       "      <td>-38</td>\n",
       "      <td>701</td>\n",
       "      <td>131</td>\n",
       "      <td>16692</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>484</td>\n",
       "      <td>61</td>\n",
       "      <td>118</td>\n",
       "      <td>39</td>\n",
       "      <td>38</td>\n",
       "      <td>-11</td>\n",
       "      <td>-116</td>\n",
       "      <td>274</td>\n",
       "      <td>245</td>\n",
       "      <td>2128</td>\n",
       "      <td>...</td>\n",
       "      <td>15691</td>\n",
       "      <td>-267</td>\n",
       "      <td>882</td>\n",
       "      <td>348</td>\n",
       "      <td>102</td>\n",
       "      <td>179</td>\n",
       "      <td>-58</td>\n",
       "      <td>533</td>\n",
       "      <td>448</td>\n",
       "      <td>15763</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>118</td>\n",
       "      <td>16</td>\n",
       "      <td>197</td>\n",
       "      <td>39</td>\n",
       "      <td>50</td>\n",
       "      <td>237</td>\n",
       "      <td>-129</td>\n",
       "      <td>311</td>\n",
       "      <td>186</td>\n",
       "      <td>1608</td>\n",
       "      <td>...</td>\n",
       "      <td>3390</td>\n",
       "      <td>-235</td>\n",
       "      <td>386</td>\n",
       "      <td>484</td>\n",
       "      <td>123</td>\n",
       "      <td>66</td>\n",
       "      <td>-45</td>\n",
       "      <td>448</td>\n",
       "      <td>106</td>\n",
       "      <td>18128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>316</td>\n",
       "      <td>0</td>\n",
       "      <td>105</td>\n",
       "      <td>104</td>\n",
       "      <td>75</td>\n",
       "      <td>-76</td>\n",
       "      <td>-383</td>\n",
       "      <td>38</td>\n",
       "      <td>261</td>\n",
       "      <td>1397</td>\n",
       "      <td>...</td>\n",
       "      <td>2621</td>\n",
       "      <td>-202</td>\n",
       "      <td>547</td>\n",
       "      <td>129</td>\n",
       "      <td>-6</td>\n",
       "      <td>121</td>\n",
       "      <td>-11</td>\n",
       "      <td>346</td>\n",
       "      <td>105</td>\n",
       "      <td>22818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>228</td>\n",
       "      <td>36</td>\n",
       "      <td>106</td>\n",
       "      <td>113</td>\n",
       "      <td>102</td>\n",
       "      <td>19</td>\n",
       "      <td>-95</td>\n",
       "      <td>141</td>\n",
       "      <td>165</td>\n",
       "      <td>1030</td>\n",
       "      <td>...</td>\n",
       "      <td>29136</td>\n",
       "      <td>-178</td>\n",
       "      <td>295</td>\n",
       "      <td>286</td>\n",
       "      <td>86</td>\n",
       "      <td>80</td>\n",
       "      <td>-32</td>\n",
       "      <td>358</td>\n",
       "      <td>65</td>\n",
       "      <td>39323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>395</td>\n",
       "      <td>73</td>\n",
       "      <td>334</td>\n",
       "      <td>73</td>\n",
       "      <td>115</td>\n",
       "      <td>417</td>\n",
       "      <td>-257</td>\n",
       "      <td>259</td>\n",
       "      <td>359</td>\n",
       "      <td>2656</td>\n",
       "      <td>...</td>\n",
       "      <td>-52</td>\n",
       "      <td>-557</td>\n",
       "      <td>641</td>\n",
       "      <td>230</td>\n",
       "      <td>147</td>\n",
       "      <td>120</td>\n",
       "      <td>30</td>\n",
       "      <td>647</td>\n",
       "      <td>13</td>\n",
       "      <td>15689</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>311</td>\n",
       "      <td>-56</td>\n",
       "      <td>235</td>\n",
       "      <td>46</td>\n",
       "      <td>59</td>\n",
       "      <td>845</td>\n",
       "      <td>-444</td>\n",
       "      <td>168</td>\n",
       "      <td>517</td>\n",
       "      <td>3570</td>\n",
       "      <td>...</td>\n",
       "      <td>252</td>\n",
       "      <td>-1295</td>\n",
       "      <td>1531</td>\n",
       "      <td>2</td>\n",
       "      <td>298</td>\n",
       "      <td>135</td>\n",
       "      <td>-116</td>\n",
       "      <td>948</td>\n",
       "      <td>181</td>\n",
       "      <td>41570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>373</td>\n",
       "      <td>-23</td>\n",
       "      <td>377</td>\n",
       "      <td>-34</td>\n",
       "      <td>-4</td>\n",
       "      <td>403</td>\n",
       "      <td>-106</td>\n",
       "      <td>85</td>\n",
       "      <td>152</td>\n",
       "      <td>1844</td>\n",
       "      <td>...</td>\n",
       "      <td>5347</td>\n",
       "      <td>-356</td>\n",
       "      <td>669</td>\n",
       "      <td>132</td>\n",
       "      <td>157</td>\n",
       "      <td>106</td>\n",
       "      <td>-50</td>\n",
       "      <td>333</td>\n",
       "      <td>240</td>\n",
       "      <td>39538</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>72 rows × 7129 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene Accession Number  A28102_at  AB000114_at  AB000115_at  AB000220_at  \\\n",
       "0                            151           72          281           36   \n",
       "1                            263           21          250           43   \n",
       "2                             88          -27          358           42   \n",
       "3                            484           61          118           39   \n",
       "4                            118           16          197           39   \n",
       "..                           ...          ...          ...          ...   \n",
       "67                           316            0          105          104   \n",
       "68                           228           36          106          113   \n",
       "69                           395           73          334           73   \n",
       "70                           311          -56          235           46   \n",
       "71                           373          -23          377          -34   \n",
       "\n",
       "Gene Accession Number  AB000381_s_at  AB000409_at  AB000410_s_at  AB000449_at  \\\n",
       "0                                 29         -299           -336           57   \n",
       "1                                  8         -103           -361          169   \n",
       "2                                 11          142           -508          359   \n",
       "3                                 38          -11           -116          274   \n",
       "4                                 50          237           -129          311   \n",
       "..                               ...          ...            ...          ...   \n",
       "67                                75          -76           -383           38   \n",
       "68                               102           19            -95          141   \n",
       "69                               115          417           -257          259   \n",
       "70                                59          845           -444          168   \n",
       "71                                -4          403           -106           85   \n",
       "\n",
       "Gene Accession Number  AB000450_at  AB000460_at  ...  Z84721_cds2_at  \\\n",
       "0                              186         1647  ...            8656   \n",
       "1                              219         2043  ...           13285   \n",
       "2                              237         1997  ...           15704   \n",
       "3                              245         2128  ...           15691   \n",
       "4                              186         1608  ...            3390   \n",
       "..                             ...          ...  ...             ...   \n",
       "67                             261         1397  ...            2621   \n",
       "68                             165         1030  ...           29136   \n",
       "69                             359         2656  ...             -52   \n",
       "70                             517         3570  ...             252   \n",
       "71                             152         1844  ...            5347   \n",
       "\n",
       "Gene Accession Number  Z84722_at  Z86000_at  Z93784_at  Z94753_s_at  \\\n",
       "0                           -156        538        434          168   \n",
       "1                           -241        863        506           70   \n",
       "2                           -581        737        770          141   \n",
       "3                           -267        882        348          102   \n",
       "4                           -235        386        484          123   \n",
       "..                           ...        ...        ...          ...   \n",
       "67                          -202        547        129           -6   \n",
       "68                          -178        295        286           86   \n",
       "69                          -557        641        230          147   \n",
       "70                         -1295       1531          2          298   \n",
       "71                          -356        669        132          157   \n",
       "\n",
       "Gene Accession Number  Z95624_at  Z96810_at  Z97054_xpt2_at  Z97074_at  \\\n",
       "0                            116       -122             398        188   \n",
       "1                             94        -40             649        147   \n",
       "2                             57        -38             701        131   \n",
       "3                            179        -58             533        448   \n",
       "4                             66        -45             448        106   \n",
       "..                           ...        ...             ...        ...   \n",
       "67                           121        -11             346        105   \n",
       "68                            80        -32             358         65   \n",
       "69                           120         30             647         13   \n",
       "70                           135       -116             948        181   \n",
       "71                           106        -50             333        240   \n",
       "\n",
       "Gene Accession Number  hum_alu_at  \n",
       "0                           15091  \n",
       "1                           11038  \n",
       "2                           16692  \n",
       "3                           15763  \n",
       "4                           18128  \n",
       "..                            ...  \n",
       "67                          22818  \n",
       "68                          39323  \n",
       "69                          15689  \n",
       "70                          41570  \n",
       "71                          39538  \n",
       "\n",
       "[72 rows x 7129 columns]"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Veuillez écrire le code ici et appeler le résultat 'total_df'\n",
    "total_df = pd.concat([train_df, test_df], axis=0, ignore_index=True)\n",
    "\n",
    "# trier les colonnes par ordre alphabétique\n",
    "total_df = total_df.sort_index(axis=1)\n",
    "\n",
    "total_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb6a6e91",
   "metadata": {},
   "source": [
    "d.  En utilisant le fichier `actual.csv`, veuillez ajouter à **total_df** une colonne intitulée **type**, qui indique pour chaque patient son type de maladie (ALL ou AML).  Attention, cette colonne servira uniquement pour l'évaluation extrinsèque en fin de labo."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "a2f70a33",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(72, 7130)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Gene Accession Number</th>\n",
       "      <th>A28102_at</th>\n",
       "      <th>AB000114_at</th>\n",
       "      <th>AB000115_at</th>\n",
       "      <th>AB000220_at</th>\n",
       "      <th>AB000381_s_at</th>\n",
       "      <th>AB000409_at</th>\n",
       "      <th>AB000410_s_at</th>\n",
       "      <th>AB000449_at</th>\n",
       "      <th>AB000450_at</th>\n",
       "      <th>AB000460_at</th>\n",
       "      <th>...</th>\n",
       "      <th>Z84722_at</th>\n",
       "      <th>Z86000_at</th>\n",
       "      <th>Z93784_at</th>\n",
       "      <th>Z94753_s_at</th>\n",
       "      <th>Z95624_at</th>\n",
       "      <th>Z96810_at</th>\n",
       "      <th>Z97054_xpt2_at</th>\n",
       "      <th>Z97074_at</th>\n",
       "      <th>hum_alu_at</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>151</td>\n",
       "      <td>72</td>\n",
       "      <td>281</td>\n",
       "      <td>36</td>\n",
       "      <td>29</td>\n",
       "      <td>-299</td>\n",
       "      <td>-336</td>\n",
       "      <td>57</td>\n",
       "      <td>186</td>\n",
       "      <td>1647</td>\n",
       "      <td>...</td>\n",
       "      <td>-156</td>\n",
       "      <td>538</td>\n",
       "      <td>434</td>\n",
       "      <td>168</td>\n",
       "      <td>116</td>\n",
       "      <td>-122</td>\n",
       "      <td>398</td>\n",
       "      <td>188</td>\n",
       "      <td>15091</td>\n",
       "      <td>ALL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>263</td>\n",
       "      <td>21</td>\n",
       "      <td>250</td>\n",
       "      <td>43</td>\n",
       "      <td>8</td>\n",
       "      <td>-103</td>\n",
       "      <td>-361</td>\n",
       "      <td>169</td>\n",
       "      <td>219</td>\n",
       "      <td>2043</td>\n",
       "      <td>...</td>\n",
       "      <td>-241</td>\n",
       "      <td>863</td>\n",
       "      <td>506</td>\n",
       "      <td>70</td>\n",
       "      <td>94</td>\n",
       "      <td>-40</td>\n",
       "      <td>649</td>\n",
       "      <td>147</td>\n",
       "      <td>11038</td>\n",
       "      <td>ALL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>88</td>\n",
       "      <td>-27</td>\n",
       "      <td>358</td>\n",
       "      <td>42</td>\n",
       "      <td>11</td>\n",
       "      <td>142</td>\n",
       "      <td>-508</td>\n",
       "      <td>359</td>\n",
       "      <td>237</td>\n",
       "      <td>1997</td>\n",
       "      <td>...</td>\n",
       "      <td>-581</td>\n",
       "      <td>737</td>\n",
       "      <td>770</td>\n",
       "      <td>141</td>\n",
       "      <td>57</td>\n",
       "      <td>-38</td>\n",
       "      <td>701</td>\n",
       "      <td>131</td>\n",
       "      <td>16692</td>\n",
       "      <td>ALL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>484</td>\n",
       "      <td>61</td>\n",
       "      <td>118</td>\n",
       "      <td>39</td>\n",
       "      <td>38</td>\n",
       "      <td>-11</td>\n",
       "      <td>-116</td>\n",
       "      <td>274</td>\n",
       "      <td>245</td>\n",
       "      <td>2128</td>\n",
       "      <td>...</td>\n",
       "      <td>-267</td>\n",
       "      <td>882</td>\n",
       "      <td>348</td>\n",
       "      <td>102</td>\n",
       "      <td>179</td>\n",
       "      <td>-58</td>\n",
       "      <td>533</td>\n",
       "      <td>448</td>\n",
       "      <td>15763</td>\n",
       "      <td>ALL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>118</td>\n",
       "      <td>16</td>\n",
       "      <td>197</td>\n",
       "      <td>39</td>\n",
       "      <td>50</td>\n",
       "      <td>237</td>\n",
       "      <td>-129</td>\n",
       "      <td>311</td>\n",
       "      <td>186</td>\n",
       "      <td>1608</td>\n",
       "      <td>...</td>\n",
       "      <td>-235</td>\n",
       "      <td>386</td>\n",
       "      <td>484</td>\n",
       "      <td>123</td>\n",
       "      <td>66</td>\n",
       "      <td>-45</td>\n",
       "      <td>448</td>\n",
       "      <td>106</td>\n",
       "      <td>18128</td>\n",
       "      <td>ALL</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 7130 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene Accession Number  A28102_at  AB000114_at  AB000115_at  AB000220_at  \\\n",
       "0                            151           72          281           36   \n",
       "1                            263           21          250           43   \n",
       "2                             88          -27          358           42   \n",
       "3                            484           61          118           39   \n",
       "4                            118           16          197           39   \n",
       "\n",
       "Gene Accession Number  AB000381_s_at  AB000409_at  AB000410_s_at  AB000449_at  \\\n",
       "0                                 29         -299           -336           57   \n",
       "1                                  8         -103           -361          169   \n",
       "2                                 11          142           -508          359   \n",
       "3                                 38          -11           -116          274   \n",
       "4                                 50          237           -129          311   \n",
       "\n",
       "Gene Accession Number  AB000450_at  AB000460_at  ...  Z84722_at  Z86000_at  \\\n",
       "0                              186         1647  ...       -156        538   \n",
       "1                              219         2043  ...       -241        863   \n",
       "2                              237         1997  ...       -581        737   \n",
       "3                              245         2128  ...       -267        882   \n",
       "4                              186         1608  ...       -235        386   \n",
       "\n",
       "Gene Accession Number  Z93784_at  Z94753_s_at  Z95624_at  Z96810_at  \\\n",
       "0                            434          168        116       -122   \n",
       "1                            506           70         94        -40   \n",
       "2                            770          141         57        -38   \n",
       "3                            348          102        179        -58   \n",
       "4                            484          123         66        -45   \n",
       "\n",
       "Gene Accession Number  Z97054_xpt2_at  Z97074_at  hum_alu_at  type  \n",
       "0                                 398        188       15091   ALL  \n",
       "1                                 649        147       11038   ALL  \n",
       "2                                 701        131       16692   ALL  \n",
       "3                                 533        448       15763   ALL  \n",
       "4                                 448        106       18128   ALL  \n",
       "\n",
       "[5 rows x 7130 columns]"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Veuillez écrire le code ici :\n",
    "total_df['type'] = actual['cancer']\n",
    "\n",
    "print(total_df.shape)\n",
    "total_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6cd4f8d8",
   "metadata": {},
   "source": [
    "e. Y a-t-il des données manquantes dans la Data Frame obtenue ?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "fad93fd0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n"
     ]
    }
   ],
   "source": [
    "# données manquantes\n",
    "print(total_df.isnull().sum().sum())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e0e8a12",
   "metadata": {},
   "source": [
    "f. Y a-t-il des *outliers* dans chaque colonne ?  Même s'il y en a, ne pas les supprimer."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5ba452b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "2f226369",
   "metadata": {},
   "source": [
    "g. Veuillez donner un résumé pour chaque colonne numérique (le _5-number-summary_) en utilisant la méthode _describe()_.  Utiliser l'affichage par défaut qui affiche un petit nombre de colonnes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c91e81d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "633ac85e",
   "metadata": {},
   "source": [
    "h. Combien de fois apparaît chaque classe de référence, ALL et AML ?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "id": "b17c18b1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Gene Accession Number</th>\n",
       "      <th>A28102_at</th>\n",
       "      <th>AB000114_at</th>\n",
       "      <th>AB000115_at</th>\n",
       "      <th>AB000220_at</th>\n",
       "      <th>AB000381_s_at</th>\n",
       "      <th>AB000409_at</th>\n",
       "      <th>AB000410_s_at</th>\n",
       "      <th>AB000449_at</th>\n",
       "      <th>AB000450_at</th>\n",
       "      <th>AB000460_at</th>\n",
       "      <th>...</th>\n",
       "      <th>Z84721_cds2_at</th>\n",
       "      <th>Z84722_at</th>\n",
       "      <th>Z86000_at</th>\n",
       "      <th>Z93784_at</th>\n",
       "      <th>Z94753_s_at</th>\n",
       "      <th>Z95624_at</th>\n",
       "      <th>Z96810_at</th>\n",
       "      <th>Z97054_xpt2_at</th>\n",
       "      <th>Z97074_at</th>\n",
       "      <th>hum_alu_at</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>151</td>\n",
       "      <td>72</td>\n",
       "      <td>281</td>\n",
       "      <td>36</td>\n",
       "      <td>29</td>\n",
       "      <td>-299</td>\n",
       "      <td>-336</td>\n",
       "      <td>57</td>\n",
       "      <td>186</td>\n",
       "      <td>1647</td>\n",
       "      <td>...</td>\n",
       "      <td>8656</td>\n",
       "      <td>-156</td>\n",
       "      <td>538</td>\n",
       "      <td>434</td>\n",
       "      <td>168</td>\n",
       "      <td>116</td>\n",
       "      <td>-122</td>\n",
       "      <td>398</td>\n",
       "      <td>188</td>\n",
       "      <td>15091</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>263</td>\n",
       "      <td>21</td>\n",
       "      <td>250</td>\n",
       "      <td>43</td>\n",
       "      <td>8</td>\n",
       "      <td>-103</td>\n",
       "      <td>-361</td>\n",
       "      <td>169</td>\n",
       "      <td>219</td>\n",
       "      <td>2043</td>\n",
       "      <td>...</td>\n",
       "      <td>13285</td>\n",
       "      <td>-241</td>\n",
       "      <td>863</td>\n",
       "      <td>506</td>\n",
       "      <td>70</td>\n",
       "      <td>94</td>\n",
       "      <td>-40</td>\n",
       "      <td>649</td>\n",
       "      <td>147</td>\n",
       "      <td>11038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>88</td>\n",
       "      <td>-27</td>\n",
       "      <td>358</td>\n",
       "      <td>42</td>\n",
       "      <td>11</td>\n",
       "      <td>142</td>\n",
       "      <td>-508</td>\n",
       "      <td>359</td>\n",
       "      <td>237</td>\n",
       "      <td>1997</td>\n",
       "      <td>...</td>\n",
       "      <td>15704</td>\n",
       "      <td>-581</td>\n",
       "      <td>737</td>\n",
       "      <td>770</td>\n",
       "      <td>141</td>\n",
       "      <td>57</td>\n",
       "      <td>-38</td>\n",
       "      <td>701</td>\n",
       "      <td>131</td>\n",
       "      <td>16692</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>484</td>\n",
       "      <td>61</td>\n",
       "      <td>118</td>\n",
       "      <td>39</td>\n",
       "      <td>38</td>\n",
       "      <td>-11</td>\n",
       "      <td>-116</td>\n",
       "      <td>274</td>\n",
       "      <td>245</td>\n",
       "      <td>2128</td>\n",
       "      <td>...</td>\n",
       "      <td>15691</td>\n",
       "      <td>-267</td>\n",
       "      <td>882</td>\n",
       "      <td>348</td>\n",
       "      <td>102</td>\n",
       "      <td>179</td>\n",
       "      <td>-58</td>\n",
       "      <td>533</td>\n",
       "      <td>448</td>\n",
       "      <td>15763</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>118</td>\n",
       "      <td>16</td>\n",
       "      <td>197</td>\n",
       "      <td>39</td>\n",
       "      <td>50</td>\n",
       "      <td>237</td>\n",
       "      <td>-129</td>\n",
       "      <td>311</td>\n",
       "      <td>186</td>\n",
       "      <td>1608</td>\n",
       "      <td>...</td>\n",
       "      <td>3390</td>\n",
       "      <td>-235</td>\n",
       "      <td>386</td>\n",
       "      <td>484</td>\n",
       "      <td>123</td>\n",
       "      <td>66</td>\n",
       "      <td>-45</td>\n",
       "      <td>448</td>\n",
       "      <td>106</td>\n",
       "      <td>18128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>316</td>\n",
       "      <td>0</td>\n",
       "      <td>105</td>\n",
       "      <td>104</td>\n",
       "      <td>75</td>\n",
       "      <td>-76</td>\n",
       "      <td>-383</td>\n",
       "      <td>38</td>\n",
       "      <td>261</td>\n",
       "      <td>1397</td>\n",
       "      <td>...</td>\n",
       "      <td>2621</td>\n",
       "      <td>-202</td>\n",
       "      <td>547</td>\n",
       "      <td>129</td>\n",
       "      <td>-6</td>\n",
       "      <td>121</td>\n",
       "      <td>-11</td>\n",
       "      <td>346</td>\n",
       "      <td>105</td>\n",
       "      <td>22818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>228</td>\n",
       "      <td>36</td>\n",
       "      <td>106</td>\n",
       "      <td>113</td>\n",
       "      <td>102</td>\n",
       "      <td>19</td>\n",
       "      <td>-95</td>\n",
       "      <td>141</td>\n",
       "      <td>165</td>\n",
       "      <td>1030</td>\n",
       "      <td>...</td>\n",
       "      <td>29136</td>\n",
       "      <td>-178</td>\n",
       "      <td>295</td>\n",
       "      <td>286</td>\n",
       "      <td>86</td>\n",
       "      <td>80</td>\n",
       "      <td>-32</td>\n",
       "      <td>358</td>\n",
       "      <td>65</td>\n",
       "      <td>39323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>395</td>\n",
       "      <td>73</td>\n",
       "      <td>334</td>\n",
       "      <td>73</td>\n",
       "      <td>115</td>\n",
       "      <td>417</td>\n",
       "      <td>-257</td>\n",
       "      <td>259</td>\n",
       "      <td>359</td>\n",
       "      <td>2656</td>\n",
       "      <td>...</td>\n",
       "      <td>-52</td>\n",
       "      <td>-557</td>\n",
       "      <td>641</td>\n",
       "      <td>230</td>\n",
       "      <td>147</td>\n",
       "      <td>120</td>\n",
       "      <td>30</td>\n",
       "      <td>647</td>\n",
       "      <td>13</td>\n",
       "      <td>15689</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>311</td>\n",
       "      <td>-56</td>\n",
       "      <td>235</td>\n",
       "      <td>46</td>\n",
       "      <td>59</td>\n",
       "      <td>845</td>\n",
       "      <td>-444</td>\n",
       "      <td>168</td>\n",
       "      <td>517</td>\n",
       "      <td>3570</td>\n",
       "      <td>...</td>\n",
       "      <td>252</td>\n",
       "      <td>-1295</td>\n",
       "      <td>1531</td>\n",
       "      <td>2</td>\n",
       "      <td>298</td>\n",
       "      <td>135</td>\n",
       "      <td>-116</td>\n",
       "      <td>948</td>\n",
       "      <td>181</td>\n",
       "      <td>41570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>373</td>\n",
       "      <td>-23</td>\n",
       "      <td>377</td>\n",
       "      <td>-34</td>\n",
       "      <td>-4</td>\n",
       "      <td>403</td>\n",
       "      <td>-106</td>\n",
       "      <td>85</td>\n",
       "      <td>152</td>\n",
       "      <td>1844</td>\n",
       "      <td>...</td>\n",
       "      <td>5347</td>\n",
       "      <td>-356</td>\n",
       "      <td>669</td>\n",
       "      <td>132</td>\n",
       "      <td>157</td>\n",
       "      <td>106</td>\n",
       "      <td>-50</td>\n",
       "      <td>333</td>\n",
       "      <td>240</td>\n",
       "      <td>39538</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>72 rows × 7129 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene Accession Number  A28102_at  AB000114_at  AB000115_at  AB000220_at  \\\n",
       "0                            151           72          281           36   \n",
       "1                            263           21          250           43   \n",
       "2                             88          -27          358           42   \n",
       "3                            484           61          118           39   \n",
       "4                            118           16          197           39   \n",
       "..                           ...          ...          ...          ...   \n",
       "67                           316            0          105          104   \n",
       "68                           228           36          106          113   \n",
       "69                           395           73          334           73   \n",
       "70                           311          -56          235           46   \n",
       "71                           373          -23          377          -34   \n",
       "\n",
       "Gene Accession Number  AB000381_s_at  AB000409_at  AB000410_s_at  AB000449_at  \\\n",
       "0                                 29         -299           -336           57   \n",
       "1                                  8         -103           -361          169   \n",
       "2                                 11          142           -508          359   \n",
       "3                                 38          -11           -116          274   \n",
       "4                                 50          237           -129          311   \n",
       "..                               ...          ...            ...          ...   \n",
       "67                                75          -76           -383           38   \n",
       "68                               102           19            -95          141   \n",
       "69                               115          417           -257          259   \n",
       "70                                59          845           -444          168   \n",
       "71                                -4          403           -106           85   \n",
       "\n",
       "Gene Accession Number  AB000450_at  AB000460_at  ...  Z84721_cds2_at  \\\n",
       "0                              186         1647  ...            8656   \n",
       "1                              219         2043  ...           13285   \n",
       "2                              237         1997  ...           15704   \n",
       "3                              245         2128  ...           15691   \n",
       "4                              186         1608  ...            3390   \n",
       "..                             ...          ...  ...             ...   \n",
       "67                             261         1397  ...            2621   \n",
       "68                             165         1030  ...           29136   \n",
       "69                             359         2656  ...             -52   \n",
       "70                             517         3570  ...             252   \n",
       "71                             152         1844  ...            5347   \n",
       "\n",
       "Gene Accession Number  Z84722_at  Z86000_at  Z93784_at  Z94753_s_at  \\\n",
       "0                           -156        538        434          168   \n",
       "1                           -241        863        506           70   \n",
       "2                           -581        737        770          141   \n",
       "3                           -267        882        348          102   \n",
       "4                           -235        386        484          123   \n",
       "..                           ...        ...        ...          ...   \n",
       "67                          -202        547        129           -6   \n",
       "68                          -178        295        286           86   \n",
       "69                          -557        641        230          147   \n",
       "70                         -1295       1531          2          298   \n",
       "71                          -356        669        132          157   \n",
       "\n",
       "Gene Accession Number  Z95624_at  Z96810_at  Z97054_xpt2_at  Z97074_at  \\\n",
       "0                            116       -122             398        188   \n",
       "1                             94        -40             649        147   \n",
       "2                             57        -38             701        131   \n",
       "3                            179        -58             533        448   \n",
       "4                             66        -45             448        106   \n",
       "..                           ...        ...             ...        ...   \n",
       "67                           121        -11             346        105   \n",
       "68                            80        -32             358         65   \n",
       "69                           120         30             647         13   \n",
       "70                           135       -116             948        181   \n",
       "71                           106        -50             333        240   \n",
       "\n",
       "Gene Accession Number  hum_alu_at  \n",
       "0                           15091  \n",
       "1                           11038  \n",
       "2                           16692  \n",
       "3                           15763  \n",
       "4                           18128  \n",
       "..                            ...  \n",
       "67                          22818  \n",
       "68                          39323  \n",
       "69                          15689  \n",
       "70                          41570  \n",
       "71                          39538  \n",
       "\n",
       "[72 rows x 7129 columns]"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total_df.iloc[:, :-1]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88ec02bb",
   "metadata": {},
   "source": [
    "i. Appliquez la PCA sur des données normalisées (sans la colonne `type`), en **deux** dimensions (sans normaliser le résultat). **Vous travaillerez sur ces données tout au long du labo.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "68703432",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[ 2.71052085e+01,  6.44848272e+00],\n",
       "       [ 3.50741799e+00,  1.58458374e+01],\n",
       "       [ 6.00063760e+01,  1.03923540e+01],\n",
       "       [ 2.06403889e+01, -2.62341392e+00],\n",
       "       [-2.10120165e+01,  4.17941334e+01],\n",
       "       [-3.45971095e+00, -1.47571630e+01],\n",
       "       [ 2.66184937e+01, -1.39258551e+01],\n",
       "       [ 5.60706238e+01, -2.21430014e+01],\n",
       "       [ 3.16280390e+01,  3.60868621e+01],\n",
       "       [-1.37259751e+01, -2.72957522e+00],\n",
       "       [-6.67705365e+00,  1.56237863e+01],\n",
       "       [-3.73402523e+01, -2.57345451e+01],\n",
       "       [-1.75212502e+01,  4.04005097e+01],\n",
       "       [ 2.14391797e+00,  1.89232501e+01],\n",
       "       [-1.93326213e+01,  4.55173136e+01],\n",
       "       [-3.77020484e+00,  1.38672093e+01],\n",
       "       [ 7.73895225e+01,  5.60597199e+01],\n",
       "       [-2.16584614e+01, -4.49393267e+00],\n",
       "       [-2.74751756e+01,  1.01059320e-01],\n",
       "       [ 2.66435181e+01,  1.15299917e+02],\n",
       "       [-4.10005569e+01,  2.06160258e+01],\n",
       "       [-1.24825420e+01, -1.97106213e+01],\n",
       "       [-4.27750011e+00, -1.72415892e+01],\n",
       "       [ 6.66826532e+00,  3.22195478e+01],\n",
       "       [-1.54217783e+01, -8.60942498e+00],\n",
       "       [-1.48917192e+01,  2.07257718e+00],\n",
       "       [ 4.28015763e+01, -2.94682759e+01],\n",
       "       [ 2.09157374e+01, -1.75205291e+01],\n",
       "       [ 7.57612596e+00, -3.24942015e+01],\n",
       "       [ 3.05191895e+01, -1.20446587e+01],\n",
       "       [ 1.98700981e+01, -1.73447205e+01],\n",
       "       [ 2.15584224e+01, -1.08861465e+01],\n",
       "       [ 2.04375823e+01, -4.24548463e+00],\n",
       "       [-4.68787179e+01, -7.04445933e+00],\n",
       "       [ 1.55535755e+01, -3.07670628e+00],\n",
       "       [ 7.26868504e+00, -2.17307509e+01],\n",
       "       [ 3.56856549e+01, -7.83695420e+00],\n",
       "       [ 7.23054956e+01, -2.31667666e+01],\n",
       "       [ 5.86992327e+00, -1.91191330e+01],\n",
       "       [ 8.42369630e+00, -2.30380187e+01],\n",
       "       [-1.30192021e+01,  4.02494721e+00],\n",
       "       [-3.24804022e+01,  5.56378134e+00],\n",
       "       [ 4.24767985e+00,  5.72231317e+01],\n",
       "       [ 6.57339226e+01, -2.29776850e+01],\n",
       "       [-4.44007723e+01,  1.33285398e+01],\n",
       "       [-3.33010032e+01, -7.03583440e+00],\n",
       "       [-4.90992801e+01, -4.86259816e+00],\n",
       "       [-4.59764111e+01, -8.11923555e+00],\n",
       "       [-4.79857668e+01, -2.91269898e+00],\n",
       "       [-4.33505699e+01, -1.70310453e+01],\n",
       "       [-2.40198262e+01, -5.08995129e-01],\n",
       "       [-7.06924725e+00,  1.06516597e+01],\n",
       "       [-1.17002049e+01,  5.57007333e+01],\n",
       "       [-1.09705547e+01,  4.79667156e+01],\n",
       "       [-2.79509944e+01,  4.18466155e+00],\n",
       "       [-2.45360947e+01, -1.44738013e+01],\n",
       "       [ 2.69171219e+00, -1.61794427e+01],\n",
       "       [-3.42724215e+01, -1.06161577e+01],\n",
       "       [-3.75396709e+01, -1.89379723e+01],\n",
       "       [ 4.28690537e+01,  1.61769154e+00],\n",
       "       [ 3.04135262e+00, -2.53736657e+01],\n",
       "       [ 1.42083065e+01, -1.76467484e+01],\n",
       "       [-1.15082618e+01,  7.24254658e+00],\n",
       "       [-3.58946490e+01, -2.48554131e+01],\n",
       "       [ 1.65782105e+01, -3.67788754e+01],\n",
       "       [-3.23102106e+01, -1.59208690e+01],\n",
       "       [ 1.34781305e+01, -1.64099982e+01],\n",
       "       [-1.55616185e+01, -2.49661609e+01],\n",
       "       [-5.53402365e+01, -2.75000091e+00],\n",
       "       [ 3.72816801e+01, -3.04533816e+00],\n",
       "       [ 9.13834163e+01, -1.41186490e+01],\n",
       "       [ 6.49193558e+00, -1.22658817e+01]])"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Veuillez nommer X_pca le résultat de la PCA sur les colonnes de données.\n",
    "# Veuillez nommer Y la colonne indiquant le type correct de chaque patient.\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "std_scaler = StandardScaler()\n",
    "scaled_df = std_scaler.fit_transform(total_df.iloc[:, :-1])\n",
    "\n",
    "from sklearn.decomposition import PCA\n",
    "pca = PCA(n_components=2)\n",
    "pca.fit_transform(scaled_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b51ff3a8",
   "metadata": {},
   "source": [
    "k. À ce stade, après la PCA, les points vous semblent-ils présenter un groupement visible ?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9242aca2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "1dcda422",
   "metadata": {},
   "source": [
    "## 2. Groupement avec _k_-moyennes\n",
    "\n",
    "Dans cette section, on vous demande de grouper les patients en utilisant la méthode des _k_-moyennes (avec initialisation _kmeans++_) et de trouver la valeur optimale du nombre de groupes entre 2 et 12, en utilisant l'indice de Davies-Bouldin et le score Silhouette.  Vous *ne* devez *pas* utiliser l'information du type de maladie.\n",
    "\n",
    "Vu le nombre élevé d'attributs, vous réutiliserez les résultat de l'ACP faite ci-dessus, sans normalisation après l'ACP.\n",
    "\n",
    "a. Pour les trois scores suivants, veuillez rappeler (en utilisant la documentation de sklearn) quel est leur intervalle de valeurs et quelles valeurs indiquent un meilleur groupement :\n",
    "* l'indice de Davies-Bouldin\n",
    "* le coefficient de Silhouette\n",
    "* l'inertie."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d535b8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Veuillez répondre ici.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "804ea67f",
   "metadata": {},
   "source": [
    "b. Pour un nombre de groupes _k_ allant de 2 à 12 compris, veuillez effectuer le groupement avec _k_-moyennes, puis représenter sur trois graphiques la variation des trois scores intrinsèques en fonction de _k_.  Utiliser `n_init=10`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "baf2cc17",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ee709be",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "c07db9f4",
   "metadata": {},
   "source": [
    "c. Selon les courbes affichées, quelle valeur de _k_ proposez-vous de retenir ?  Veuillez donner une raison.\n",
    "\n",
    "Si vous exécutez plusieurs fois l'affichage des courbes, la valeur optimale de _k_ reste-t-elle la même ?  Que pouvez-vous en conclure ?\n",
    "\n",
    "Note : si vous fixez l'attribut `random_seed` de KMeans, alors les courbes ne varient pas au fil des exécutions, sauf si vous changez `n_init`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1af2d844",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Votre discussion ici."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab59eac8",
   "metadata": {},
   "source": [
    "d. Avec la valeur de _k_ choisie au (c), veuillez calculer deux scores extrinsèques : l'indice de Rand ajusté et le score F1.  Veuillez commenter brièvement ces scores.\n",
    "\n",
    "_Indication pour calculer le F1-score, qui n'est pas implémenté dans sklearn pour le groupement._  Pour chaque patient, assignez-lui le type qui est majoritaire dans son groupe (ALL ou AML) en utilisant les bonnes réponses.  Puis comparez les types ainsi assignés avec les types corrects en utilisant directement le score F1 de sklearn.  Notez que pour simplifier, cette méthode est un peu différente de celle du cours."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de1f45c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Effectuer le groupement k-moyennes avec la meilleure valeur de k retenue au (b).\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ca204d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calcul et affichage des deux scores.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0158931",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Votre commentaire sur les scores."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "09253aa7",
   "metadata": {},
   "source": [
    "e. Veuillez représenter les points avec PCA en 2D, en gardant les marqueurs précédents pour les deux types (cercle et triangle), et en utilisant des couleurs pour coder les groupes trouvés. Suggestion : réutiliser le code du Labo 1.  Veuillez commenter le résultat."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "708d0bbb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Utiliser le groupement obtenu ci-dessus.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2263bfed",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Veuillez écrire ici votre discussion des résultats, à l'aide du graphique.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e24d4d81",
   "metadata": {},
   "source": [
    "## 3. Groupement avec DBSCAN\n",
    "\n",
    "Dans cette section, vous répondrez aux mêmes questions que dans la section précédente, mais pour la méthode DBSCAN, en utilisant les mêmes mesures intrinsèques que ci-dessus (sauf l'inertie).\n",
    "\n",
    "a. Veuillez représenter sur des _heatmaps_ la variation des deux scores intrinsèques et le nombre de groupes en fonction des paramètres _eps_ et _min_samples_ :\n",
    "* chercher _eps_ entre 0.5 et 14 par sauts de 0.5\n",
    "* chercher _min_samples_ entre 1 et 12"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "3f97d66b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2, 7129)"
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Obtenir les deux scores et le nombre de groupes en variant les deux paramètres, les mettre dans des tableaux 2D\n",
    "pca.components_.shape\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b242ecf9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Afficher les trois heatmaps : Davies_Bouldin, Silhouette, et Nombre_de_groupes\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aba082d2",
   "metadata": {},
   "source": [
    "b. Selon les scores trouvés, quelles valeurs de _eps_ et _min_samples_ proposez-vous de retenir ?  Pourquoi ?\n",
    "\n",
    "Attention : les trois heatmaps ne sont pas toujours en accord, donc vous avez une certaine liberté pour choisir les deux paramètres (il n'y a pas d'optimum évident)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f06a273",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Votre discussion ici."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56def74e",
   "metadata": {},
   "source": [
    "c. Avec les valeurs choisies au (b), veuillez calculer les deux scores extrinsèques comme dans la partie (2)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f10910d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Effectuer le groupement DBSCAN avec les meilleurs paramètres,\n",
    "# puis calculer l'indice de Rand ajusté et le score F1.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3be33296",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77081225",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Votre commentaire sur les scores."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f87169e",
   "metadata": {},
   "source": [
    "d. Veuillez représenter les points en 2D avec la même représentation que ci-dessus : utilisez des marques différentes pour les deux types, et des couleurs pour coder les groupes trouvés.  Veuillez commenter le résultat."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b1f6eb8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3803566e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Veuillez commenter le graphique ici.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "527c6de5",
   "metadata": {},
   "source": [
    "## 4. Conclusion\n",
    "\n",
    "Comment se comparent les deux méthodes optimisées, en termes de scores extrinsèques ?  Veuillez discuter les paramètres trouvés et les scores obtenus.  En particulier, y a-t-il des sous-types des leucémies ALL et AML qui sont visibles sur les groupements que vous obtenus ?  Combien pour chacun ?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39077d2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Votre discussion ici."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e19be49",
   "metadata": {},
   "source": [
    "***\n",
    "**Fin du Labo 2.**  Veuillez nettoyer ce notebook en gardant seulement les réponses et résultats désirés, l'enregistrer en remplaçant 'student' par votre nom, et le soumettre sur Cyberlearn."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
